src/fu-common-cab.c

```diff
@@ -0,0 +1,327 @@
+/* -*- Mode: C; tab-width: 8; indent-tabs-mode: t; c-basic-offset: 8 -*-
+ *
+ * Copyright (C) 2017 Richard Hughes <richard@hughsie.com>
+ *
+ * Licensed under the GNU Lesser General Public License Version 2.1
+ *
+ * This library is free software; you can redistribute it and/or
+ * modify it under the terms of the GNU Lesser General Public
+ * License as published by the Free Software Foundation; either
+ * version 2.1 of the License, or (at your option) any later version.
+ *
+ * This library is distributed in the hope that it will be useful,
+ * but WITHOUT ANY WARRANTY; without even the implied warranty of
+ * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the GNU
+ * Lesser General Public License for more details.
+ *
+ * You should have received a copy of the GNU Lesser General Public
+ * License along with this library; if not, write to the Free Software
+ * Foundation, Inc., 51 Franklin Street, Fifth Floor, Boston, MA  02110-1301 USA
+ */
+
+#include "config.h"
+
+#include <libgcab.h>
+
+#include "fu-common-cab.h"
+#include "fu-common.h"
+
+#include "fwupd-error.h"
+
+#ifdef HAVE_GCAB_1_0
+
+/* sets the firmware and signature blobs on AsRelease */
+static gboolean
+fu_common_store_from_cab_release (AsRelease *release, GCabFolder *cabfolder, GError **error)
+{
+	AsChecksum *csum_tmp;
+	GCabFile *cabfile;
+	GBytes *blob;
+	guint64 size;
+	g_autofree gchar *basename = NULL;
+	g_autofree gchar *checksum = NULL;
+	const gchar *suffixes[] = { "asc", "p7b", "p7c", NULL };
+
+	/* ensure we always have a content checksum */
+	csum_tmp = as_release_get_checksum_by_target (release, AS_CHECKSUM_TARGET_CONTENT);
+	if (csum_tmp == NULL) {
+		g_autoptr(AsChecksum) csum = as_checksum_new ();
+		as_checksum_set_target (csum, AS_CHECKSUM_TARGET_CONTENT);
+		/* if this isn't true, a firmware needs to set in
+		 * the metainfo.xml file something like:
+		 * <checksum target="content" filename="FLASH.ROM"/> */
+		as_checksum_set_filename (csum, "firmware.bin");
+		as_release_add_checksum (release, csum);
+		csum_tmp = csum;
```

Comment:
Don't you wanna steal the pointer here, because of the `g_autoptr()` above?

Reply:
No, because as_release_add_checksum adds a reference, so we're still "borrowing" the ref from AsRelease like we get from as_release_get_checksum_by_target(). I agree it's inelegant.

Reply:
maybe we just call 	`csum_tmp = as_release_get_checksum_by_target();` again?

Reply:
Ah, gotcha, make sense.